base_wamas/models/wamas_document_element.py

```diff
@@ -0,0 +1,129 @@
+# Copyright 2023 Jacques-Etienne Baudoux (BCIM) <je@bcim.be>
+# Copyright 2023 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from collections import OrderedDict
+
+from odoo import _, fields, models
+from odoo.exceptions import UserError
+
+from ..models.wamas_document_template import check_type_of_item
+
+
+def get_item(recordset_field, row, idx_start, res, convert_type):
+    for field in recordset_field:
+        if field._name == "wamas.document.default.field.template.line":
+            field = field.df_field_id
+
+        idx_end = idx_start + field.len_field
+
+        val = row[idx_start:idx_end]
+        if convert_type:
+            val = field.get_value_from_string(val)
+
+        if isinstance(res, OrderedDict):
+            res[field.code] = val
+        elif isinstance(res, list):
+            res.append((field.code, val))
+
+        idx_start = idx_end
+
+    return res, idx_start
+
+
+class WamasDocumentElement(models.Model):
+    _name = "wamas.document.element"
+    _description = "WAMAS Document Element"
+
+    name = fields.Char(required=True)
+    code = fields.Char(required=True)
+    df_field_template_id = fields.Many2one(
+        "wamas.document.default.field.template", "Default Field Template", required=True
+    )
+    field_ids = fields.One2many("wamas.document.field", "element_id", "Fields")
+
+    def get_values_from_element(
+        self, row, convert_type=False, result_type="ordered_dict"
+    ):
+        self.ensure_one()
+
+        res = False
+
+        if result_type == "list_tuple":
```

Comment:
What is the purpose of having a different result?
IMO it would be better to have 2 methods...

Reply:
> What is the purpose of having a different result?
- the `list_tuple`: `list` is each line of UBL document and each `tuple` (in `list`) is each element in the line -> one `list_tuple` related to one line -> it's so generic to use for any purpose
- the `ordered_dict`: `ordered_dict` is a line and each item of `ordered_dict` is each element in the line -> I use this to build the `ordered dict` structure for converting wamas -> ubl in `base_wamas_ubl`

> IMO it would be better to have 2 methods...
I agree, it should be better if have 2 methods, I'll split it into 2 methods